statistics_service/native/src/impl/StatisticsService.cpp

```diff
@@ -1,164 +1,234 @@
-
 #include "StatisticsService.hpp"
-#include <utils/Log.h>
+
+#include <algorithm>
 #include <fstream>
+#include <numeric>
 #include <sstream>
 
+#include <utils/Log.h>
+
 #ifdef LOG_TAG
 #undef LOG_TAG
 #endif
 #define LOG_TAG "statistics_service"
 
 namespace statistics {
-    namespace service {
-
-        ::std::shared_ptr<StatisticsService> StatisticsService::S_INSTANCE = NULL;
-
-        ::std::shared_ptr<StatisticsService> StatisticsService::getInstance()
-        {
-            if (S_INSTANCE == NULL)
-            {
-                S_INSTANCE = ndk::SharedRefBase::make<StatisticsService>();
-            }
-            return S_INSTANCE;
-        }
+namespace service {
 
-        StatisticsService::current_line_to_read = 1;
+std::shared_ptr<StatisticsService> StatisticsService::instance_ = nullptr;
 
-        StatisticsService::StatisticsService()
-        {
+std::string StatisticsService::GetServiceName() {
+    return std::string() + descriptor + "/default";
+}
 
-        }
+std::shared_ptr<StatisticsService> StatisticsService::GetInstance() {
+    if (instance_ == nullptr)

```

Comment:
Aceasta varianta de singleton poate produce un race condition atunci cand 2 thread-uri executa functia getInstance in paralel.
Putem evita aceasta problema mutand initializarea lui S_INSTANCE in function getInstance ca "static function variable". Atunci, vom avea garantia ca variabla va fi initializata o singura data: https://stackoverflow.com/a/8102145
```suggestion
    static std::shared_ptr<StatisticsService> s_Instance =  SharedRefBase::make<StatisticsService>();
    return s_Instance;
```